File: apps/kafka/workers/download_worker.py
import base64
import io
import multiprocessing
import threading
import time
from workers.kafka_utils import Consumer, Producer
from workers.common import load_yaml_config, print_counter
from loguru import logger
from PIL import Image
import aiohttp
import asyncio
import torch
import torch.multiprocessing as mp
from torch.utils.data import IterableDataset, DataLoader


async def download_image(session, url):
    try:
        async with session.get(url) as response:
            image_data = await response.read()
            image_bytes = base64.b64encode(image_data)
            return image_bytes.decode('utf-8')
    except Exception as e:
        logger.error(f"Error downloading image from {url}: {e}")
        return None

async def download_images(urls):
    async with aiohttp.ClientSession() as session:
        tasks = [download_image(session, url) for url in urls]
        return await asyncio.gather(*tasks)


class ImageUrlDataset(IterableDataset):
    def __init__(self, consumer_topic, rank):
        self.consumer_topic = consumer_topic
        self.rank = rank

    def __iter__(self):
        consumer = Consumer(self.consumer_topic, partition_id=self.rank)
        for message in consumer.consumer:
            message_data = message.value

            image_contents = asyncio.run(download_images(message_data["image_urls"]))

            for img, doc_id in zip(image_contents, message_data["document_ids"]):
                if img is None:
                    logger.warning(f"Skipping document {doc_id} due to image download error")
                    continue
                yield {
                    "image_contents": img,
                    "document_ids": doc_id
                }


class DownloadWorker:
    STAGE = "download_images"

    def __init__(self, config, rank, counter):
        self.stage_config = config["stages"][self.STAGE]
        logger.info(f"Config: {self.stage_config}")
        
        consumer_topic = self.stage_config["consumer"]

        producer_topics = self.stage_config["producer_list"]
        self.producers = [
            Producer(
                producer_topic, 
                config["kafka_topics"][producer_topic]["partitions"]
            ) for producer_topic in producer_topics
        ]

        self.dataset = ImageUrlDataset(consumer_topic, rank)
        batch_size = self.stage_config["batch_size"]
        self.dataloader = DataLoader(self.dataset, batch_size=batch_size, num_workers=1, pin_memory=True, prefetch_factor=3)
        
        self.counter = counter

    def run(self):
        for idx, batch in enumerate(self.dataloader):
            # logger.info(f"Processing batch {idx} of {len(batch['document_ids'])} documents")
            with self.counter.get_lock():
                self.counter.value += len(batch['document_ids'])
            
            message = {
                "image_contents": [img for img in batch["image_contents"]],
                "document_ids": batch["document_ids"].tolist() if torch.is_tensor(batch["document_ids"]) else batch["document_ids"]
            }
            total_size = sum(len(img) for img in message["image_contents"])
            logger.info(f"Batch {idx} total image size: {total_size/1024/1024:.2f} MB")
            
            for producer in self.producers:
                producer.send(idx, message)


def download_image_worker(rank, config, counter):
    worker = DownloadWorker(config, rank, counter)
    worker.run()


if __name__ == "__main__":
    mp.set_start_method('spawn', force=True)

    config = load_yaml_config("configs/example_config.yaml")

    topic = config["stages"]["download_images"]["consumer"]
    N_PROCESSES = config["kafka_topics"][topic]["partitions"]

    counter = print_counter()

    mp.spawn(
        download_image_worker,
        args=(config, counter),
        nprocs=N_PROCESSES,
        join=True
    )

    # with multiprocessing.Pool(N_PROCESSES) as pool:
    #     pool.starmap(download_image, [(i, config) for i in range(N_PROCESSES)])

File: apps/kafka/workers/download_worker.go
package main

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"

	"sync"
	"time"

	"sync/atomic"

	"github.com/IBM/sarama"
)

const (
	TIMEOUT = 80 * time.Second
)

var counter atomic.Uint32

func main() {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Load config
	config := load_yaml_config("configs/pd12m_config.yaml")

	// Configure Sarama
	saramaConfig := sarama.NewConfig()
	saramaConfig.Producer.Return.Successes = false
	saramaConfig.Producer.Return.Errors = false
	saramaConfig.Producer.RequiredAcks = sarama.NoResponse
	saramaConfig.Producer.Compression = sarama.CompressionGZIP
	saramaConfig.Producer.MaxMessageBytes = 1000000000 // 1GB
	saramaConfig.Version = sarama.V0_11_0_2

	// Create sync producer
	producer, err := sarama.NewAsyncProducer([]string{"localhost:9092"}, saramaConfig)
	if err != nil {
		log.Fatal("Failed to create producer:", err)
	}
	defer producer.Close()

	// Create consumer
	consumer, err := sarama.NewConsumer([]string{"localhost:9092"}, saramaConfig)
	if err != nil {
		log.Fatal("Failed to create consumer:", err)
	}
	defer consumer.Close()

	consumerTopic := config.Stages["download_images"].Consumer
	producerTopic := config.Stages["download_images"].ProducerList[0]

	// Create worker pool
	numConsumerPartitions := config.KafkaTopics[consumerTopic].Partitions
	// numConsumerPartitions := 1
	producerPartitions := config.KafkaTopics[producerTopic].Partitions
	var wg sync.WaitGroup

	producerConfig := ProducerConfig{
		Name:       producerTopic,
		Partitions: producerPartitions,
		Producer:   producer,
	}

	// Track the number of messages processed
	docsChan := make(chan int, 1024)

	log.Printf("Producer Partitions: %d, Consumer Partitions: %d", producerPartitions, numConsumerPartitions)
	// Start a worker for each partition
	for partition := 0; partition < numConsumerPartitions; partition++ {
		wg.Add(1)
		go func(partition int32) {
			defer wg.Done()
			partitionConsumer, err := consumer.ConsumePartition(
				consumerTopic,
				partition,
				sarama.OffsetNewest,
			)
			if err != nil {
				log.Printf("Failed to create partition consumer for partition %d: %v", partition, err)
				return
			}
			defer partitionConsumer.Close()
			worker(ctx, partitionConsumer, partition, producerConfig, docsChan)
		}(int32(partition))
	}

	go processDocument(ctx, docsChan)

	// Wait for all workers to complete
	wg.Wait()
}

func worker(ctx context.Context, consumer sarama.PartitionConsumer, consumerPartition int32,
	producerConfig ProducerConfig, docsChan chan int) {
	msgChan := make(chan *sarama.ConsumerMessage, 1024)

	// Read messages from Kafka in parallel and push to channel
	go func() {
		for {
			select {
			case msg := <-consumer.Messages():
				msgChan <- msg
			case <-ctx.Done():
				close(msgChan)
				return
			}
		}
	}()

	for {
		select {
		case msg := <-msgChan:
			// Process message here
			var data MessageContent
			if err := json.Unmarshal(msg.Value, &data); err != nil {
				log.Printf("Error unmarshaling JSON: %v", err)
				continue
			}

			downloadAndSend(ctx, consumerPartition, producerConfig, data, docsChan)
		case <-ctx.Done():
			return
		}
	}
}

func processDocument(ctx context.Context, docsChan chan int) {

	totalDocuments := 0
	currProcessedDocuments := 0
	totalSize := 0
	// Create a ticker to print processed documents every minute
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case size := <-docsChan:
			totalDocuments++
			currProcessedDocuments++
			totalSize += size
		case <-ticker.C:
			// Print and reset the counter for the current interval
			fmt.Printf("\n---------------------------------------------------------------------------------------------\n")
			fmt.Printf("Processed %d documents/second of size %s in the last 10 seconds. Total processed: %d\n",
				currProcessedDocuments/10, humanReadableSize(totalSize), totalDocuments)
			fmt.Printf("---------------------------------------------------------------------------------------------")
			totalSize = 0
			currProcessedDocuments = 0
		case <-ctx.Done():
			fmt.Println("Stopping document processing...")
			return
		}
	}
}

func downloadAndSend(ctx context.Context, consumerPartition int32, producerConfig ProducerConfig, data MessageContent, docsChan chan int) {
	// start := time.Now()
	results := downloadImages(ctx, producerConfig, data, docsChan)

	// successful := 0
	// failed := 0

	// totalSize := 0
	// Process results
	docsToWait := int(float64(len(data.ImageURLs)) * 0.1)
	for i := 0; i < docsToWait; i++ {
		<-results
		// result := <-results
		// if result.Success {
		// 	successful++
		// 	totalSize += result.Size
		// } else {
		// 	failed++
		// 	// log.Printf("Failed to download %s: %v", result.URL, result.Error)
		// }
	}

	// close(results)
	// duration := time.Since(start)
	// log.Printf("Rank %d: Downloaded %d images (%d successful, %d failed) in %.2f seconds, downloaded %s\n",
	// 	consumerPartition, len(data.ImageURLs), successful, failed, duration.Seconds(), humanReadableSize(totalSize))
}

func downloadImages(ctx context.Context, producerConfig ProducerConfig, data MessageContent, docsChan chan int) chan DownloadResult {
	numImages := len(data.ImageURLs)
	results := make(chan DownloadResult, numImages)
	jobs := make(chan Job, numImages)

	// Start workers
	for i := 0; i < numImages; i++ {
		go func() {
			for job := range jobs {
				select {
				case results <- downloadImage(ctx, producerConfig, job.DocumentID, job.URL, docsChan):
				case <-ctx.Done():
					return
				}
			}
		}()
	}

	// Send jobs
	go func() {
		for i := range data.ImageURLs {
			job := Job{
				URL:        data.ImageURLs[i],
				DocumentID: data.DocumentIDs[i],
			}

			select {
			case jobs <- job:
			case <-ctx.Done():
				break
			}
		}
		close(jobs)
	}()

	return results
}

func downloadImage(ctx context.Context, producerConfig ProducerConfig, documentID string, url string, docsChan chan int) DownloadResult {
	client := &http.Client{Timeout: TIMEOUT}

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return DownloadResult{URL: url, Success: false, Error: err}
	}

	resp, err := client.Do(req)
	if err != nil {
		return DownloadResult{URL: url, Success: false, Error: err}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return DownloadResult{URL: url, Success: false, Error: fmt.Errorf("HTTP %d", resp.StatusCode)}
	}

	// Use a streaming encoder to directly encode base64 while reading the image
	var base64Buffer bytes.Buffer
	base64Writer := base64.NewEncoder(base64.StdEncoding, &base64Buffer)
	if _, err := io.Copy(base64Writer, resp.Body); err != nil {
		return DownloadResult{URL: url, Success: false, Error: err}
	}
	base64Writer.Close() // Ensure the base64 encoder flushes any remaining data

	payload := MessagePayload{
		DocumentIDs:   []string{documentID},
		ImageContents: []string{base64Buffer.String()}, // Avoid additional memory allocations
	}

	jsonBuffer := &bytes.Buffer{}
	encoder := json.NewEncoder(jsonBuffer)
	if err := encoder.Encode(payload); err != nil {
		return DownloadResult{URL: url, Success: false, Error: err}
	}

	partitionNumber := int(counter.Add(1)) % producerConfig.Partitions

	producerConfig.Producer.Input() <- &sarama.ProducerMessage{
		Topic:     producerConfig.Name,
		Key:       sarama.StringEncoder(fmt.Sprintf("partition-%d", partitionNumber)),
		Value:     sarama.ByteEncoder(jsonBuffer.Bytes()),
		Partition: int32(partitionNumber),
	}

	docsChan <- jsonBuffer.Len()
	return DownloadResult{URL: url, Success: true, Error: nil, Size: jsonBuffer.Len()}
}
